docs(android): Introduce Android ProfilingManager options (#18481)
getsentry/sentry-docs: docs/platforms/android/profiling/legacy/index.mdx

---
title: Legacy Profiling
sidebar_title: Legacy Profiling

> description: "Learn about the legacy profiling implementations and how to use them on devices below Android 15."

Reviewer: ```suggestion
description: "Learn about the legacy profilers and how to use them on devices running Android versions earlier than Android 15."
```

sidebar_order: 8000
---

<Alert level="warning" title="Legacy">
  This page documents Sentry's legacy Android profilers. The recommended way to

>   profile your app is the{" "}

Reviewer: ```suggestion
  profile your app is the is the <PlatformLink to="/profiling/">ProfilingManager (Perfetto) based UI Profiling</PlatformLink>, available since SDK version `8.45.0`. Use the legacy profilers only when you need to support devices running below Android 15 (API level 35), or while migrating from an older setup.
```

Reviewer: Putting this all on one line to fix a formatting issue.

>   <PlatformLink to="/profiling/">

Reviewer: ```suggestion
```

>     ProfilingManager (Perfetto) based UI Profiling

Reviewer: ```suggestion
```

>   </PlatformLink>

Reviewer: ```suggestion
```

>   , available since SDK version `8.45.0`. Use the legacy profilers only when you

Reviewer: ```suggestion
```

>   need to support devices running below Android 15 (API level 35), or while

Reviewer: ```suggestion
```

>   migrating from an older setup.

Reviewer: ```suggestion
```

</Alert>

There are two legacy profiling implementations:

- **Legacy UI Profiling**, which samples threads using the Android runtime's `tracer`. It supports the same `manual` and `trace` lifecycle modes as the current UI Profiling, but uses the `tracer` backend instead of the ProfilingManager.
- **Transaction-based profiling**, the oldest implementation, which runs only alongside performance transactions and is limited to 30 seconds.

<Alert level="warning" title="Important">
  Legacy UI Profiling and transaction-based profiling use the Android runtime's
  `tracer` under the hood to sample threads. There are known issues that this

>   `tracer` can cause crashes in certain circumstances. See this{" "}

Reviewer: ```suggestion
  `tracer` can cause crashes in certain circumstances. See this <PlatformLink to="/profiling/troubleshooting#i-see-an-elevated-number-of-crashes-in-the-android-runtime-when-profiling-is-activated"> troubleshooting </PlatformLink> entry for more information. The ProfilingManager (Perfetto) based <PlatformLink to="/profiling/">UI Profiling</PlatformLink> is not affected by these issues.
```

>   <PlatformLink to="/profiling/troubleshooting#i-see-an-elevated-number-of-crashes-in-the-android-runtime-when-profiling-is-activated">

Reviewer: ```suggestion
```

>     troubleshooting

Reviewer: ```suggestion
```

>   </PlatformLink>{" "}

Reviewer: ```suggestion
```

>   entry for more information. The ProfilingManager (Perfetto) based{" "}

Reviewer: ```suggestion
```

>   <PlatformLink to="/profiling/">UI Profiling</PlatformLink> is not affected by

Reviewer: ```suggestion
```

>   these issues.

Reviewer: ```suggestion
```

</Alert>

## Legacy UI Profiling

> Legacy UI Profiling is available starting in SDK version `8.7.0` and is supported on API level 21 and up.

Reviewer: ```suggestion
Legacy UI Profiling is available starting with SDK version `8.7.0` and is supported on API level 21 and up.
```

    Sentry.stopProfiler();
  }
}
```

```kotlin
import io.sentry.ProfileLifecycle
import io.sentry.Sentry
import io.sentry.android.core.SentryAndroid

// App main Application class
class MyApplication : Application() {

  override fun onCreate() {
     super.onCreate()
     SentryAndroid.init(
        this,
        { options ->
          options.dsn = "___PUBLIC_DSN___"
          // Enable UI profiling, adjust in production env. This is evaluated only once per session
          options.profileSessionSampleRate = 1.0
          options.profileLifecycle = ProfileLifecycle.MANUAL
          // Enable profiling on app start. The app start profile has to be stopped through Sentry.stopProfiler()
          options.isStartProfilerOnAppStart = true
         })
    // Start profiling, if lifecycle is set to `manual` and startProfilerOnAppStart is set to `true`
    Sentry.startProfiler()
    // Stop profiling, if lifecycle is set to `manual` and startProfilerOnAppStart is set to `true`.
    // This call is optional. If you don't stop the profiler, it will keep profiling your application until the process exits.
    Sentry.stopProfiler()
  }
}
```

## Transaction-Based Profiling

<Alert>

> This is the oldest profiling implementation and will eventually be deprecated. We recommend upgrading to the <PlatformLink to="/profiling/">ProfilingManager (Perfetto) based UI Profiling</PlatformLink>. The same behaviour, without the 30 seconds limitation, can be achieved with the trace lifecycle. In order to upgrade, you also need to remove the <PlatformLink to="/configuration/options/#transaction-based-profiling-options">transaction-based profiling options</PlatformLink> from your configuration.

Reviewer: ```suggestion
This is the oldest profiling implementation and will eventually be deprecated. We recommend upgrading to the <PlatformLink to="/profiling/">ProfilingManager (Perfetto) based UI Profiling</PlatformLink>. The same behavior, without the 30 seconds limitation, can be achieved with the trace lifecycle. In order to upgrade, you also need to remove the <PlatformLink to="/configuration/options/#transaction-based-profiling-options">transaction-based profiling options</PlatformLink> from your configuration.
```
